services/login_sso.go

```diff
@@ -0,0 +1,236 @@
+
+	ssoReq := models.SSORequest{
+		LicenseKey: u.LicenseKey,
+	}
+
+	if ssoReq.LicenseKey == "" {
+		return nil, errors.New("missing license key")
+	}
+
+	b, err := json.Marshal(ssoReq)
+	if err != nil {
+		return nil, errors.New("failed to marshal payload")
+	}
+
+	rUrl := ssoUrl + ssoRedirectPath
+	req, err := http.NewRequest("POST", rUrl, bytes.NewBuffer(b))
+	if err != nil {
+		return nil, errors.New("failed to create SSO request")
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{
+		Timeout: 15 * time.Second,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("failed to call SSO: %+v", err)
+		return nil, errors.New("failed to call SSO")
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.New("failed to read SSO response")
+	}
+
+	var ssoResp models.SSOResponse
+	if err := json.Unmarshal(body, &ssoResp); err != nil {
+		return nil, errors.New("failed to parse SSO JSON response")
+	}
+
+	if !ssoResp.Status || ssoResp.Data.RedirectURL == "" {
+		log.Errorf("no valid redirectUrl in SSO response: %+v", ssoResp)
+		return nil, errors.New("no redirect URL in SSO response")
+	}
+
+	log.Infof("Login should be successful: %+v", ssoResp)
+
+	return &models.SSOLoginResponse{
+		RedirectURL: ssoResp.Data.RedirectURL,
+	}, nil
+}
+
+func (u *LoginUserSSOService) RedeemToken(queryValues url.Values) (*models.SSOTokenResponse, error) {
+	accessCode := queryValues.Get("saml_access_code")
+
+	if accessCode == "" {
+		return nil, errors.New("missing saml_access_code")
+	}
+
+	tokenRequest := models.SSOTokenRequest{
+		Token: accessCode,
+	}
+
+	b, err := json.Marshal(tokenRequest)
+	if err != nil {
+		return nil, errors.New("failed to marshal payload")
+	}
+
+	rUrl := ssoUrl + ssoTokenPath
+	req, err := http.NewRequest("POST", rUrl, bytes.NewBuffer(b))
+	if err != nil {
+		return nil, errors.New("failed to create SSO token request")
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{
+		Timeout: 15 * time.Second,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("failed to call SSO: %+v", err)
+		return nil, errors.New("failed to call SSO for token")
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.New("failed to read SSO token response")
+	}
+
+	var tokenResponse models.SSOTokenResponse
+	if err := json.Unmarshal(body, &tokenResponse); err != nil {
+		return nil, errors.New("failed to parse SSO token JSON response")
+	}
+
+	if !tokenResponse.Status {
+		log.Errorf("failed to redeem token in response: %+v", tokenResponse)
+		return nil, errors.New("failed to redeem token")
+	}
+
+	if tokenResponse.Data.Payload.Email == "" {
+		return nil, errors.New("no email in token response")
+	}
+	if tokenResponse.Data.Payload.OrganizationExternalID == "" {
+		return nil, errors.New("no email in token response")
```

Comment:
I'm not sure if this error message is on purpose?

Reply:
It's not. An oversight that will be corrected